argoverse/map_representation/map_api_v2.py

```diff
@@ -0,0 +1,463 @@
+        """Retrieve the two principal edges of the crosswalk, in 2d."""
+        return (self.edge1[:, :2], self.edge2[:, :2])
+
+    def __eq__(self, other: "PedCrossing") -> bool:
+        """Check if two pedestrian crossing objects are equal, up to a tolerance."""
+        return np.allclose(self.edge1, other.edge1) and np.allclose(self.edge2, other.edge2)
+
+
+class LocalLaneMarking(NamedTuple):
+    """Information about a lane marking, representing either the left or right boundary of a lane segment.
+
+    Args:
+        bound_type: type of marking that represents the lane boundary, e.g. "SOLID_WHITE" or "DASHED_YELLOW".
+        src_lane_id: id of lane segment to which this lane marking belongs.
+        bound_side: string representing which side of a lane segment this marking represents, i.e. "left" or "right".
+        polyline: array of shape (N,3) representing the waypoints of the lane segment's marked boundary.
+    """
+
+    bound_type: str
+    src_lane_id: int
+    bound_side: str
+    polyline: np.ndarray
+
+
+@dataclass(frozen=False)
+class VectorLaneSegment:
+    """
+    Args:
+        id: unique identifier for this lane segment (guaranteed to be unique only within this local map).
+        right_ln_bound: array of shape (M,3) representing the right lane boundary.
+        right_bound_type: type of marking that represents the right lane boundary.
+        r_neighbor_id: unique identifier of the lane segment representing this object's right neighbor.
+        left_ln_bound: array of shape (N,3) representing the right lane boundary
+        left_bound_type: type of marking that represents the left lane boundary.
+        l_neighbor_id: unique identifier of the lane segment representing this object's left neighbor.
+        predecessors: unique identifiers of lane segments that are predecessors of this object.
+        successors: unique identifiers of lane segments that represent successor of this object.
+        lane_type:
+        polygon_boundary: array of shape (N,3) 
+        is_intersection: boolean value representing whether or not this lane segment lies within an intersection.
+        render_l_bound: boolean flag for visualization, indicating whether to render the left lane boundary.
+        render_r_bound: boolean flag for visualization, indicating whether to render the right lane boundary.
+    """
+
+    id: Optional[int] = None
+    right_ln_bound: Optional[np.ndarray] = None
+    right_bound_type: Optional[str] = None
+    r_neighbor_id: Optional[int] = None
+    left_ln_bound: Optional[np.ndarray] = None
+    left_bound_type: Optional[str] = None
+    l_neighbor_id: Optional[int] = None
+    predecessors: Optional[List[int]] = None
+    successors: Optional[List[int]] = None
+    lane_type: Optional[str] = None
+    polygon_boundary: Optional[np.ndarray] = None
+    is_intersection: Optional[bool] = None
+    render_l_bound: Optional[bool] = True
+    render_r_bound: Optional[bool] = True
```

Comment:
Are all of the optional values necessary? I generally prefer default values for initialization for types that are "traditionally" primitives. My reasoning is three-fold:

1. Removes complexity.
2. Remove possibility of calling a method on `None`.
3. Makes development with `mypy` easier as we won't have to constantly check for `None` values.

What do you think?

Reply:
Hmm so you are thinking to just keep all of the attributes and types above, except remove the Optional prefix?

So for many lanes, there is no left or right neighbor, or no successor, so `None` is a requirement there. So I don't think we can get away with removing them all. Or, we could make an Enum for every thing, but I'm not sure how that could replace a list or interoperate with a list of neighbors

Reply:
> "traditionally" primitives

Yeah, I meant for things such as `int`, `str`, `bool`, etc.